group_theory/perm.lean

```diff
@@ -6,11 +6,132 @@ Authors: Chris Hughes
+
+lemma mem_of_mem_perms_of_list : ∀ {l : list α} {f : perm α}, f ∈ perms_of_list l → ∀ {x}, f x ≠ x → x ∈ l
+| []     f h := have f = 1 := by simpa [perms_of_list] using h, by rw this; simp
+| (a::l) f h :=
+(mem_append.1 h).elim
+  (λ h x hx, mem_cons_of_mem _ (mem_of_mem_perms_of_list h hx))
+  (λ h x hx,
+    let ⟨y, hy, hy'⟩ := mem_bind.1 h in
+    let ⟨g, hg₁, hg₂⟩ := mem_map.1 hy' in
+    if hxa : x = a then by simp [hxa]
+    else if hxy : x = y then mem_cons_of_mem _ $ by rwa hxy
+    else mem_cons_of_mem _ $
+    mem_of_mem_perms_of_list hg₁ $
+      by rw [eq_inv_mul_iff_mul_eq.2 hg₂, mul_apply, swap_inv, swap_apply_def];
+        split_ifs; cc)
+
+lemma mem_perms_of_list_iff {l : list α} {f : perm α} : f ∈ perms_of_list l ↔ ∀ {x}, f x ≠ x → x ∈ l :=
+⟨mem_of_mem_perms_of_list, mem_perms_of_list_of_mem⟩
+
+lemma nodup_perms_of_list : ∀ {l : list α} (hl : l.nodup), (perms_of_list l).nodup
+| []     hl := by simp [perms_of_list]
+| (a::l) hl :=
+have hl' : l.nodup, from nodup_of_nodup_cons hl,
+have hln' : (perms_of_list l).nodup, from nodup_perms_of_list hl',
+have hmeml : ∀ {f : perm α}, f ∈ perms_of_list l → f a = a,
+  from λ f hf, not_not.1 (mt (mem_of_mem_perms_of_list hf) (nodup_cons.1 hl).1),
+by rw [perms_of_list, list.nodup_append, list.nodup_bind, pairwise_iff_nth_le]; exact
+⟨hln', ⟨λ _ _, nodup_map (λ _ _, (mul_left_inj _).1) hln',
+  λ i j hj hij x hx₁ hx₂,
+    let ⟨f, hf⟩ := mem_map.1 hx₁ in
+    let ⟨g, hg⟩ := mem_map.1 hx₂ in
+    have hix : x a = nth_le l i (lt_trans hij hj),
+      by rw [← hf.2, mul_apply, hmeml hf.1, swap_apply_left],
+    have hiy : x a = nth_le l j hj,
+      by rw [← hg.2, mul_apply, hmeml hg.1, swap_apply_left],
+    absurd (hf.2.trans (hg.2.symm)) $
+      λ h, ne_of_lt hij $ nodup_iff_nth_le_inj.1 hl' i j (lt_trans hij hj) hj $
+        by rw [← hix, hiy]⟩,
+  λ f hf₁ hf₂,
+    let ⟨x, hx, hx'⟩ := mem_bind.1 hf₂ in
+    let ⟨g, hg⟩ := mem_map.1 hx' in
+    have hgxa : g⁻¹ x = a, from f.bijective.1 $
+      by rw [hmeml hf₁, ← hg.2]; simp,
+    have hxa : x ≠ a, from λ h, (list.nodup_cons.1 hl).1 (h ▸ hx),
+    (list.nodup_cons.1 hl).1 $
+      hgxa ▸ mem_of_mem_perms_of_list hg.1 (by rwa [apply_inv_self, hgxa])⟩
+
+def perms_of_finset (s : finset α) : finset (perm α) :=
+quotient.hrec_on s.1 (λ l hl, ⟨perms_of_list l, nodup_perms_of_list hl⟩)
+  (λ a b hab, hfunext (congr_arg _ (quotient.sound hab))
+    (λ ha hb _, heq_of_eq $ finset.ext.2 $
+      by simp [mem_perms_of_list_iff,mem_of_perm hab]))
+s.2
+
+lemma mem_perms_of_finset_iff : ∀ {s : finset α} {f : perm α},
+  f ∈ perms_of_finset s ↔ ∀ {x}, f x ≠ x → x ∈ s :=
+begin
+  rintros ⟨s, hs⟩ f,
+  cases quotient.exists_rep s with l hl,
+  subst hl,
+  exact mem_perms_of_list_iff
+end
+
+lemma card_perms_of_finset : ∀ (s : finset α),
+  (perms_of_finset s).card = s.card.fact :=
+begin
+  rintros ⟨s, hs⟩,
+  cases quotient.exists_rep s with l hl,
```

Comment:
you can merge this now in one line:
```lean
  rintros ⟨⟨l, rfl⟩, hl⟩,
```
also in the previous proof

Reply:
Surprisingly this worked. I'm impressed. `quotient.induction_on` didn't work because it couldn't compute the motive, hence the use of `subst`